src/synth/keyboard_sdl.ml

```diff
@@ -77,37 +82,41 @@ object (self)
     let m = m.Frame.midi in
     let t =
       let ans = MIDI.create (MFrame.size ()) in
-        Sdlevent.pump ();
-        while Sdlevent.has_event () do
-          try
-            match Sdlevent.poll () with
-              | Some (Sdlevent.KEYDOWN k) ->
-                  let c = Sdlkey.char_of_key k.Sdlevent.keysym in
-                    if c = '+' || c = '*' then
-                      velocity <- min 1. (velocity +. 0.1)
-                    else if c = '-' || c = '/' then
-                      velocity <- max 0. (velocity -. 0.1)
-                    else
-                      let n = note_of_char c in
-                        (* Printf.printf "Playing note %d.\n%!" n; *)
-                        MIDI.insert ans (0,MIDI.Note_on (n, velocity))
-              | Some (Sdlevent.KEYUP k) ->
-                  let c = Sdlkey.char_of_key k.Sdlevent.keysym in
-                  let n = note_of_char c in
-                    (* Printf.printf "Stopping note %d.\n%!" n; *)
-                    MIDI.insert ans (0,MIDI.Note_off (n, velocity))
-              | _ -> ()
-          with
-            | Not_found
-            | Invalid_argument _ -> ()
-        done;
-        ans
-    in
-      for c = 0 to Array.length m - 1 do
-        MIDI.clear_all m.(c);
-        MIDI.merge m.(c) t
+      Sdl.pump_events ();
+      while Sdl.has_event Sdl.Event.key_down || Sdl.has_event Sdl.Event.key_up do
```

Comment:
This is a long imperative loop. It could probably benefit from being factored out as a tail-rec function.